### PR TITLE
Add a block fixtures framework to allow automated checking of block validation

### PR DESCRIPTION
**Pull request by glendaviesnz**
**Note** Needs to be merged into https://github.com/Automattic/jetpack/pull/18582  before it is merged.
 
Adds automated checking of block validation for Jetpack blocks. See p1HpG7-b6y-p2 for background.

Based on the [Gutenberg core block fixtures](https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/fixtures/blocks/README.md), but restructured to put the fixtures next to the block in question.

Only runs for WhatsApp block currently.

To generate new fixtures:

`yarn fixtures:generate extensions/blocks/send-a-message/whatsapp-button/test/validate.js`

To run validation tests:

`yarn test-extensions-single extensions/blocks/send-a-message/whatsapp-button/test/validate.js`

See the [readme file](https://github.com/Automattic/jetpack/pull/18643/files#diff-90609b3338a3a938e3b869f04e4fadfe4e6a8f36cf30dc328edf852964869a1a) for more details.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/shared/test/block-fixtures.js
Comment (aaronrobertshaw) [inline projects/plugins/jetpack/extensions/shared/test/block-fixtures.js:275]: I don't think this is being used anywhere. Maybe I missed something but if not, perhaps we should remove it?
```suggestion
```

**Review by aaronrobertshaw — COMMENTED** [review]
This tests well for me 👍 

* All tests passed when run as per instructions
* Altering the fixture or the save function correctly caused test failures
* Altering the number of deprecations was also caught
* Deleting fixture files was handled
* Regeneration of fixtures worked and tests were passing again
* Took it for a test run on the `jetpack/rating-star` block and appeared fine
* Only one minor issue in code, inline comment left.

It would be good to get a few extra sets of eyes on this though with regards to overall approach. In the meantime, I have a couple of questions or thoughts on it.

#### Naming convention for "settings"
In https://github.com/Automattic/jetpack/pull/18582 the block's controls get refactored out into a `settings.js` file. Now within this PR we are importing block settings as `settings` within `validate.js`. It might be good to establish some clearer naming conventions around this now at the outset.

#### Automating script/fixture generation
Can we automate the creation of the initial fixture `.html` file at all? 

It seems like the simple steps of inserting a block with blank or default attributes and running it through the save function could be doable. Perhaps we could use the block settings' `example.attributes` if available?

Similar question for the `validate.js` can we script it's initial creation?

I'm thinking making it super simple to get the files in place might help encourage inclusion of these tests for more blocks.

**Comment by glendaviesnz** [comment]
>  Naming convention for "settings"
> In #18582 the block's controls get refactored out into a `settings.js` file.

Good point  - in #18582 I have renamed the toolbar and sidebar component to 'Configuration' to avoid any confusion

File: projects/plugins/jetpack/extensions/shared/test/block-fixtures.js
Comment (glendaviesnz) [reply projects/plugins/jetpack/extensions/shared/test/block-fixtures.js]: Removed

**Comment by glendaviesnz** [comment]
> Automating script/fixture generation
> Can we automate the creation of the initial fixture `.html` file at all?
> > Similar question for the `validate.js` can we script it's initial creation?

Could be worth looking at. I will get some more feedback on the concept in general first and then take a look at that.

**Review by creativecoder — COMMENTED** [review]
Have only skimmed through this, for now, but ❤️ adding this kind of test!

It'd be great to add a section to the block fixtures readme with a step by step process for updating when creating a deprecation... I imagine that's going to be "first contact" with this system for many folks working on existing blocks.

File: projects/plugins/jetpack/package.json
Comment (brbrr) [inline projects/plugins/jetpack/package.json:173]: I wonder why do we need this dependency. I was not been able to figure what parts of it are used here.

File: projects/plugins/jetpack/package.json
Comment (brbrr) [inline projects/plugins/jetpack/package.json:183]: Do we really need it? I see a bunch of yarn scripts that passing env variables as part of the script command, and I don't think there were reports of any problems with that. Cc @jeherve for a broader perspective.

File: projects/plugins/jetpack/extensions/shared/test/block-fixtures.md
Comment (brbrr) [inline projects/plugins/jetpack/extensions/shared/test/block-fixtures.md:1]: Can we also add a section on how to run these tests?

**Comment by brbrr** [comment]
<3 it! 

I left few questions inline, but I also have a few more general questions.

- Is there a way to run _all_ fixture validation tests with a single command? It could be useful when we would be adding it into our CI setup.  
- Do you plan to add these new tests into CI?

File: projects/plugins/jetpack/package.json
Comment (glendaviesnz) [reply projects/plugins/jetpack/package.json]: I was initially hoping to use some of the fixture utility code from the gutenberg test package but it wasn't exported in a way that was useful, but forgot to remove the dependency - gone now.

File: projects/plugins/jetpack/package.json
Comment (glendaviesnz) [reply projects/plugins/jetpack/package.json]: This was a copy-paste from how the gutenberg fixtures were set up, but as you note there are other instances of env variables passed without this, so have removed for now.

**Comment by glendaviesnz** [comment]
> It'd be great to add a section to the block fixtures readme with a step by step process for updating when creating a deprecation

Have added some details, let me know if it make sense or not.

File: projects/plugins/jetpack/extensions/shared/test/block-fixtures.md
Comment (glendaviesnz) [reply projects/plugins/jetpack/extensions/shared/test/block-fixtures.md:1]: Done.

**Comment by glendaviesnz** [comment]
> * Is there a way to run _all_ fixture validation tests with a single command? It could be useful when we would be adding it into our CI setup.

The validation tests all get run as part of `yarn test-extensions` test script, but have added a separate script `fixtures:test` to allow running just the validation tests.

> * Do you plan to add these new tests into CI?

I am assuming that because they all run as part of `test-extensions` they will be pulled into CI by default? In gutenberg they just run as part of the jest unit test run, do you think this is appropriate in this context or should they be excluded from unit tests runs and run as a separate set of tests?

**Comment by brbrr** [comment]
> The validation tests all get run as part of yarn test-extensions test script

Oh, I didn't know that. that should take care of CI run. Happy to approve it once you'll get a green light from one of your teammates

**Review by apeatling — APPROVED** [review]
Great work on this! Works well for me, and happy to see this implemented in new and existing blocks going forward. 👍🏻

**Review by brbrr — APPROVED** [review]
I guess you don't need approval from Crew to merge this branch into another (non-master) branch, but here it goes anyway :) 


:shipit:

**Pull request merged**
